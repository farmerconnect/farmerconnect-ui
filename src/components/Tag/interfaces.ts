import { HTMLAttributes } from 'react';
import CustomStyles from '../CustomStyles';

export enum TagVariantEnum {
  DANGER = 'danger',
  PRIMARY = 'primary',
  SUCCESS = 'success',
  WARNING = 'warning',
  SECONDARY = 'secondary',
}

export type TagVariant =
  | TagVariantEnum.DANGER
  | TagVariantEnum.PRIMARY
  | TagVariantEnum.SUCCESS
  | TagVariantEnum.WARNING
  | TagVariantEnum.SECONDARY;

interface ITagColor {
  color: string;
  backgroundColor: string;
}

export interface ITag {
  colors?: {
    [TagVariantEnum.DANGER]?: ITagColor;
    [TagVariantEnum.PRIMARY]?: ITagColor;
    [TagVariantEnum.SUCCESS]?: ITagColor;
    [TagVariantEnum.WARNING]?: ITagColor;
    [TagVariantEnum.SECONDARY]?: ITagColor;
  };
}

export interface ITagProps extends CustomStyles, HTMLAttributes<HTMLDivElement> {
  color?: string;
  variant?: TagVariant;
  backgroundColor?: string;
}
