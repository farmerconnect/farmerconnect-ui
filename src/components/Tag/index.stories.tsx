import { Story, Meta } from '@storybook/react/types-6-0';
import Tag from './';
import { TagVariantEnum } from './interfaces';

export default {
  title: 'Tag',
  component: Tag,
  argTypes: {
    color: { control: 'color' },
    variant: {
      options: Object.values(TagVariantEnum),
      control: { type: 'select' },
    },
    customStyles: { control: 'object' },
    backgroundColor: { control: 'color' },
  },
  decorators: [(Story) => <div style={{ display: 'flex' }}>{Story()}</div>],
} as Meta;

const Template: Story<any> = (args) => <Tag {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  variant: 'primary',
  children: 'Tag',
};

export const Secondary = Template.bind({});
Secondary.args = {
  variant: 'secondary',
  children: 'Tag',
};

export const Success = Template.bind({});
Success.args = {
  variant: 'success',
  children: 'Tag',
};

export const Danger = Template.bind({});
Danger.args = {
  variant: 'danger',
  children: 'Tag',
};

export const Warning = Template.bind({});
Warning.args = {
  variant: 'warning',
  children: 'Tag',
};

export const CustomColors = Template.bind({});
CustomColors.args = {
  color: 'cyan',
  backgroundColor: 'darkcyan',
  children: 'Tag',
};

export const MultipleTags = () => (
  <>
    <Tag variant="primary">Tag 1</Tag>
    <Tag variant="danger">Tag 2</Tag>
  </>
);
