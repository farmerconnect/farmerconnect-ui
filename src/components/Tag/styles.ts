import styled, { css } from 'styled-components';
import { ITagProps, TagVariant, TagVariantEnum } from './interfaces';

export const tagDefaultStyles = {
  colors: {
    [TagVariantEnum.PRIMARY]: {
      color: '#081851',
      backgroundColor: '#D1EBEE',
    },
    [TagVariantEnum.SECONDARY]: {
      color: '#5C5C5C',
      backgroundColor: '#EFEFEF',
    },
    [TagVariantEnum.SUCCESS]: {
      color: '#005E3A',
      backgroundColor: '#CEE9DD',
    },
    [TagVariantEnum.DANGER]: {
      color: '#FB2E4C',
      backgroundColor: '#FFEAED',
    },
    [TagVariantEnum.WARNING]: {
      color: '#AE8800',
      backgroundColor: '#FFFCD4',
    },
  },
};

interface IGetColor {
  theme: any;
  color?: string;
  variant?: TagVariant;
  backgroundColor?: string;
}

interface IGetColorByKey extends IGetColor {
  key: 'color' | 'backgroundColor';
}

const getColorByKey = ({ theme, variant, key }: IGetColorByKey): string => {
  if (!variant) return tagDefaultStyles.colors.primary[key];

  return theme?.tag?.colors[variant][key] || tagDefaultStyles.colors[variant][key];
};

const getTextColor = ({ theme, variant, color }: IGetColor): string =>
  color || getColorByKey({ theme, variant, key: 'color' });

const getBackgroundColor = ({ theme, variant, backgroundColor }: IGetColor): string =>
  backgroundColor || getColorByKey({ theme, variant, key: 'backgroundColor' });

export const Container = styled.div<ITagProps>`
  border-radius: 12px;
  font-size: 11px;
  font-weight: bold;
  letter-spacing: 0.05em;
  line-height: 15px;
  padding: 5px 12px;
  text-align: center;
  text-transform: uppercase;

  ${({ color, variant, backgroundColor, theme }) => css`
    color: ${getTextColor({ theme, variant, color })};
    background-color: ${getBackgroundColor({ theme, variant, backgroundColor })};
  `}

  & + & {
    margin-left: 8px;
  }
`;
