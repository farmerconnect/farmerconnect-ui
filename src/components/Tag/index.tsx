import React from 'react';
import { ITagProps, TagVariantEnum } from './interfaces';
import { Container } from './styles';

const Tag: React.FC<ITagProps> = ({ children, variant, customStyles, ...props }) => (
  <Container variant={variant} customStyles={customStyles} {...props}>
    {children}
  </Container>
);

Tag.defaultProps = {
  variant: TagVariantEnum.PRIMARY,
};

export default Tag;
