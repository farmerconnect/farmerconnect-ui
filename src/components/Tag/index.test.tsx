import { shallow } from 'enzyme';
import Tag from './';

describe('Tag component', () => {
  it('should match the snapshot', () => {
    const wrapper = shallow(<Tag>Test</Tag>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should call the function correctly', () => {
    const fn = jest.fn();
    const wrapper = shallow(<Tag onClick={fn}>Test</Tag>);

    wrapper.simulate('click');

    expect(fn).toHaveBeenCalled();
  });
});
