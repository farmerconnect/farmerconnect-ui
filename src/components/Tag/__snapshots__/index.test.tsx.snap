// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag component should match the snapshot 1`] = `
<styled.div
  variant="primary"
>
  Test
</styled.div>
`;
